pyls/rpc_manager.py

```diff
@@ -0,0 +1,190 @@
+# Copyright 2017 Palantir Technologies, Inc.
+import logging
+from uuid import uuid4
+
+from concurrent.futures import ThreadPoolExecutor, Future
+from jsonrpc.base import JSONRPCBaseResponse
+from jsonrpc.jsonrpc1 import JSONRPC10Response
+from jsonrpc.jsonrpc2 import JSONRPC20Response, JSONRPC20Request
+from jsonrpc.exceptions import JSONRPCMethodNotFound, JSONRPCDispatchException, JSONRPCServerError
+
+log = logging.getLogger(__name__)
+
+RESPONSE_CLASS_MAP = {
+    "1.0": JSONRPC10Response,
+    "2.0": JSONRPC20Response
+}
+
+
+class MissingMethodException(Exception):
+    pass
+
+
+class JSONRPCManager(object):
+
+    def __init__(self, message_manager, message_handler):
+        self._message_manager = message_manager
+        self._message_handler = message_handler
+        self._shutdown = False
+        self._sent_requests = {}
+        self._received_requests = {}
+        self._executor_service = ThreadPoolExecutor(max_workers=5)
+
+    def start(self):
+        """Start reading JSONRPC messages off of rx"""
+        self.consume_requests()
+
+    def shutdown(self):
+        """Set flag to ignore all non exit messages"""
+        self._shutdown = True
+
+    def exit(self):
+        """Stop listening for new message"""
+        self._executor_service.shutdown()
+        self._message_manager.close()
+
+    def call(self, method, params=None):
+        """Send a JSONRPC request.
+
+        Args:
+            method (str): The method name of the message to send
+            params (dict): The payload of the message
+
+        Returns:
+            Future that will resolve once a response has been received
+
+        """
+        log.debug('Calling %s %s', method, params)
+        request = JSONRPC20Request(_id=uuid4().int, method=method, params=params)
+        request_future = Future()
+        self._sent_requests[request._id] = request_future
+        self._message_manager.write_message(request)
+        return request_future
+
+    def notify(self, method, params=None):
+        """Send a JSONRPC notification.
+
+         Args:
+             method (str): The method name of the notification to send
+             params (dict): The payload of the notification
+         """
+        log.debug('Notify %s %s', method, params)
+        notification = JSONRPC20Request(method=method, params=params, is_notification=True)
+        self._message_manager.write_message(notification)
+
+    def cancel(self, request_id):
+        """Cancel pending request handler.
+
+        Args:
+            request_id (string | number): The id of the original request
+
+        Note:
+            Request will only be cancelled if it has not begun execution.
+        """
+        log.debug('Cancel request %d', request_id)
+        try:
+            self._received_requests[request_id].cancel()
+        except KeyError:
+            log.error('Received cancel for finished/nonexistent request %d', request_id)
```

Comment:
I don't think this is always an error. If I receive a "cancelRequest" for a method that didn't implement itself asynchronously then it's expected behaviour. Shouldn't be `log.error`.

Reply:
I believe that the message is completely accurate. happy to make it a debug i suppose